lib/vmdb/settings/database_source.rb

```diff
@@ -3,23 +3,51 @@ class Settings
     class DatabaseSource
       include Vmdb::Logging
 
-      attr_reader :resource
+      attr_reader :resource, :settings_change_holder
 
-      def initialize(resource)
-        @resource = resource
+      def initialize(resource_to_get_settings_about, class_to_call_settings_changes_on)
+        @resource = resource_to_get_settings_about
+        @settings_change_holder = class_to_call_settings_changes_on
+      end
+
+      def self.sources_for(resource_to_get_settings_about)
+        return [] if resource_to_get_settings_about.nil?
+        hierarchy_index = SETTINGS_HIERARCHY.index(resource_to_get_settings_about.class.name.to_sym)
+        SETTINGS_HIERARCHY[0..hierarchy_index].collect do |class_to_call_settings_changes_on|
+          new(resource_to_get_settings_about, class_to_call_settings_changes_on)
+        end
+      end
+
+      def self.parent_sources_for(resource_to_get_settings_about)
+        return [] if resource_to_get_settings_about.nil?
+        hierarchy_index = SETTINGS_HIERARCHY.index(resource_to_get_settings_about.class.name.to_sym)
+        SETTINGS_HIERARCHY[0..hierarchy_index][0...-1].collect do |class_to_call_settings_changes_on|
+          new(resource_to_get_settings_about, class_to_call_settings_changes_on)
+        end
       end
 
       def load
         return if resource.nil?
-
-        resource.settings_changes.reload.each_with_object({}) do |c, h|
+        obj_with_settings = settings_holder
+        obj_with_settings.settings_changes.reload.each_with_object({}) do |c, h|
           h.store_path(c.key_path, c.value)
-        end
+        end unless obj_with_settings.nil?
```

Comment:
i was trying to avoid calling settings_holder twice since it is reloading resource